test/init_test.go

```diff
@@ -0,0 +1,183 @@
+// +build e2e
+
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains initialization logic for the tests, such as special magical global state that needs to be initialized.
+
+package test
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+
+	knativetest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/tektoncd/pipeline/pkg/names"
+	"golang.org/x/xerrors"
+	yaml "gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+var initMetrics sync.Once
+
+func setup(t *testing.T) (*clients, string) {
+	t.Helper()
+	namespace := names.SimpleNameGenerator.RestrictLengthWithRandomSuffix("arrakis")
+
+	initializeLogsAndMetrics(t)
+
+	c := newClients(t, knativetest.Flags.Kubeconfig, knativetest.Flags.Cluster)
+	createNamespace(t, namespace, c.KubeClient)
+	verifyDefaultServiceAccountExists(t, namespace, c.KubeClient)
+	return c, namespace
+}
+
+func tearDown(t *testing.T, cs *clients, namespace string) {
+	t.Helper()
+	if cs.KubeClient == nil {
+		return
+	}
+	if t.Failed() {
+		header(t.Logf, fmt.Sprintf("Dumping objects from %s", namespace))
+		bs, err := getCRDYaml(cs, namespace)
+		if err != nil {
+			t.Error(err)
+		} else {
+			t.Log(string(bs))
+		}
+	}
+
+	t.Logf("Deleting namespace %s", namespace)
+	if err := cs.KubeClient.CoreV1().Namespaces().Delete(namespace, &metav1.DeleteOptions{}); err != nil {
+		t.Errorf("Failed to delete namespace %s: %s", namespace, err)
+	}
+}
+
+func header(logf logging.FormatLogger, text string) {
+	left := "### "
+	right := " ###"
+	txt := left + text + right
+	bar := strings.Repeat("#", len(txt))
+	logf(bar)
+	logf(txt)
+	logf(bar)
+}
+
+func initializeLogsAndMetrics(t *testing.T) {
+	t.Helper()
+	initMetrics.Do(func() {
+		flag.Parse()
+		if err := flag.Set("alsologtostderr", "true"); err != nil {
+			t.Fatalf("Failed to set 'alsologtostderr' flag to 'true': %s", err)
+		}
+		logging.InitializeLogger(knativetest.Flags.LogVerbose)
+
+		if knativetest.Flags.EmitMetrics {
+			logging.InitializeMetricExporter(t.Name())
+		}
+	})
+}
+
+func createNamespace(t *testing.T, namespace string, kubeClient kubernetes.Interface) {
+	t.Helper()
+	t.Logf("Create namespace %s to deploy to", namespace)
```

Comment:
maybe this is something you can tell me from running the tests yourself: im under the impression that the `t.Log` functions buffer their output and therefore won't show up until a test is done (or some random flush point during the test). If that's the case, then i think using `log` is better b/c folks running these tests can see incremental progress (since they can be slow)

(This same issue exists in pipelines and is on my sad secret TODO list to fix :(((( - if you can confirm this behaviour i pledge to make a real issue in the pipeline backlog about it asap XD)

Reply:
So, from what I understand, not using `t.Logf` can mess up where your logs get printed. So while they might be logged in "real time," the logs might look jumbled (some info here https://stackoverflow.com/questions/23205419/how-do-you-print-in-a-go-test-using-the-testing-package).

There is an approved proposal open to stream `t.Log` output. Looks interesting, although I don't know if/when it will be available: https://github.com/golang/go/issues/24929.